lib/stripe/stripe_configuration.rb

```diff
@@ -0,0 +1,171 @@
+# frozen_string_literal: true
+
+module Stripe
+  # Configurable options:
+  #
+  # =ca_bundle_path=
+  # The location of a file containing a bundle of CA certificates. By default
+  # the library will use an included bundle that can successfully validate
+  # Stripe certificates.
+  #
+  # =log_level=
+  # When set prompts the library to log some extra information to $stdout and
+  # $stderr about what it's doing. For example, it'll produce information about
+  # requests, responses, and errors that are received. Valid log levels are
+  # `debug` and `info`, with `debug` being a little more verbose in places.
+  #
+  # Use of this configuration is only useful when `.logger` is _not_ set. When
+  # it is, the decision what levels to print is entirely deferred to the logger.
+  #
+  # =logger=
+  # The logger should support the same interface as the `Logger` class that's
+  # part of Ruby's standard library (hint, anything in `Rails.logger` will
+  # likely be suitable).
+  #
+  # If `.logger` is set, the value of `.log_level` is ignored. The decision on
+  # what levels to print is entirely deferred to the logger.
+  class StripeConfiguration
+    DEFAULT_CA_BUNDLE_PATH = File.expand_path("../data/ca-certificates.crt",
+                                              __dir__)
+
+    # map to the same values as the standard library's logger
+    LEVEL_DEBUG = Logger::DEBUG
+    LEVEL_ERROR = Logger::ERROR
+    LEVEL_INFO = Logger::INFO
+
+    attr_accessor :api_key
+    attr_accessor :enable_telemetry
+    attr_accessor :logger
+    attr_accessor :stripe_account
+
+    attr_reader :api_base
+    attr_reader :ca_bundle_path
+    attr_reader :connect_base
+    attr_reader :initial_network_retry_delay
+    attr_reader :log_level
+    attr_reader :max_network_retries
+    attr_reader :max_network_retry_delay
+    attr_reader :open_timeout
+    attr_reader :proxy
+    attr_reader :read_timeout
+    attr_reader :uploads_base
+    attr_reader :verify_ssl_certs
+
+    def self.setup
+      new.tap do |instance|
+        yield(instance) if block_given?
+      end
+    end
+
+    # Create a new config based off an existing one. This is useful when the
+    # caller wants to override the global configuration
+    def reverse_duplicate_merge(hash)
+      dup.tap do |instance|
+        hash.each do |option, value|
+          instance.public_send("#{option}=", value)
+        end
+      end
+    end
```

Comment:
This is what I intend to use in #921 for the client configuration.

I went back and forth on whether or not the duplicate configuration object should be a snapshot or a pointer with overrides (e.g. any options not set would continually point to the global config object in the event those change). However, I feel like that could lead to unexpected behavior, which is why I favored a snapshot.